### ISSUE TITLE
NCDHW = {1,640,320,320,100} => "src/tensor.cpp:68: Invalid strides. Strides must be greater than 0"

### ISSUE THREAD

**Issue by atamazov** [ROCm/MIOpen#847]
develop at 0155ee381.
```bash
$ MIOPEN_DEBUG_FIND_ONLY_SOLVER=GemmFwd1x1_0_1_int8 ./bin/MIOpenDriver convint8 --spatial_dim 3 -I NCDHW -O NCDHW -f NCDHW -n 1 -c 640 -H 320 -W 320 -k 1 -y 1 -x
 1 -p 0 -q 0 -u 1 -v 1 --in_d 100 --fil_d 1 --pad_d 0 --conv_stride_d 1 -F 1 -V 1 -i 1 $MIO -V 0
MIOpenDriver convint8 --spatial_dim 3 -I NCDHW -O NCDHW -f NCDHW -n 1 -c 640 -H 320 -W 320 -k 1 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 --in_d 100 --fil_d 1 --pad_d 0 --conv_stride_d 1 -F 1 -V 1 -i 1 -C /spb-hlc-fs2/1/users/atamaz
ov/miopen-verify2 -V 0
MIOpen(HIP): Info [get_device_name] Raw device name: gfx906
MIOpen(HIP): Info [Handle] stream: 0x28a0ca0, device_id: 0
MIOpen(HIP): Info [GetFindModeValueImpl] MIOPEN_FIND_MODE = NORMAL(1)
MIOpen Error: /home/atamazov/github/MLOpen1/src/tensor.cpp:68: Invalid strides. Strides must be greater than 0.
MIOpen Error: /home/atamazov/github/MLOpen1/src/convolution.cpp:154: Types do not match for the filter
terminate called after throwing an instance of 'miopen::Exception'
  what():  /home/atamazov/github/MLOpen1/driver/tensor_driver.hpp:124: unmatched layout and dimension size
Aborted (core dumped)
```
If you change `--in_d`  from 100 to 10, then error goes away. Most likely, the reason is that strides are computed as 32 bit integers. 
```
640*320*320*100=6553600000 > INT_MAX - FAILS
640*320*320*10=655360000 < INT_MAX - OK
```

**Comment by carlushuang** [comment]
The root cause is we use `int` type in API level to feed in the dimension/stride information for tensor descriptor, like [`TensorDescriptor::TensorDescriptor()`](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/tensor.cpp#L60), or [`miopenSetTensorDescriptor()`](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/tensor_api.cpp#L122). To solve this problem, the best way I can image is to develop a new set of tensor API ( *Ex like `miopenSetTensorDescriptorEx()`)

However, see [cudnn's API doc](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnSetTensorNdDescriptor), they still use `int` as input arg, even when the *Ex API.

need your feedback @pfultz2 @jerryyin @sunway513

**Comment by sunway513** [comment]
cc @deven-amd  @jithunnair-amd and @jeffdaily  for comments..

**Comment by jerryyin** [comment]
Agreed with @carlushuang 

I see this Note from the cudnn API site:

> Note: The total size of a tensor including the potential padding between dimensions is limited to 2 Giga-elements of type datatype.

It looks like either 6553600000 or 655360000 exceed the 2 Giga elements limit so will just return the `CUDNN_STATUS_NOT_SUPPORTED` in cudnn. We might need an additional return status for this.

**Comment by atamazov** [comment]
@carlushuang 
> However, see [cudnn's API doc](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnSetTensorNdDescriptor), they still use `int` as input arg, even when the *Ex API.

This is because they explicitly specify the format by means of https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnTensorFormat_t.

**Comment by carlushuang** [comment]
Thanks all and here a summary of solutions according to above information

1. **limit a tensor size to 2 Giga elements**. If bigger than this size simply return `CUDNN_STATUS_NOT_SUPPORTED`. @jerryyin 
  **pros**: same behavior as cudnn. very small change is needed based on current MIOpen implementation
  **cons**: we just can't deal with tensor bigger than 2 Giga elements, while indeed internally we can treat big tensor.
2. **Develop a new set of API** and accept dimension/stride of `size_t` type @carlushuang 
  **pros**: internal/external API is now aligned. we can indeed treat big tensor the same way we treat small tensor
  **cons**: have compatible issue. 1st not compatible with convention of cudnn, 2nd if stride is come from a pointer, you need use a value of `size_t` not `int`. 
3. **parse in tensor format information** while create tensor descriptor. @atamazov 
  **pros**: we only need parse in the value of each dimension, not the stride. since the stride is the most possible to exceed 2G, so now we only need limit each dimention to be with in 2G, and support range is still very big
  **cons**: violate previous discussion in https://github.com/ROCmSoftwarePlatform/MIOpen/issues/767, https://github.com/ROCmSoftwarePlatform/MIOpen/pull/343. still have the problem if user use API that need to parse in stride.

since solution#1 is the simplest one, so @sunway513 @deven-amd  please help check if solution#2, solution#3 is acceptable in framework's point of view. Beside, if we really go solution#1, we just drop the support of >2G tensor? or is there new approach we can support this.

cc @junliume

**Comment by deven-amd** [comment]
TF actually already calculates dims/strides in `int64` and converts them to `int` in-order to accomodate MIOpen API. So switching to using `int64` for dims/strides should be not be an issue for TF

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/stream_executor/rocm/rocm_dnn.cc#L662-L676

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/stream_executor/rocm/rocm_dnn.cc#L756-L773

**Comment by carlushuang** [comment]
OK. In this case, per @atamazov 's suggestion, we can have a *V2 API to deal with the big tensor support (solution_2) by promoting stride/dimention to `size_t`, while let current API return _unsupport_ (solution_1)

```
extern "C" miopenStatus_t miopenSetTensorDescriptorV2(miopenTensorDescriptor_t tensorDesc,
    miopenDataType_t dataType,
    size_t nbDims,
    size_t* dimsA,
    size_t* stridesA);
```
this function can be make default when switching to semi-hypothetic Find 2.0 API

new tensor format support together with this V2 API is yet to be discussed.

cc @asroy @JehandadKhan

**Comment by atamazov** [comment]
Correction:
- In the V2 prototype, _my recommendation_ is changing the only type of `*stridesA` to `size_t`.

> new tensor format support together with this V2 API is yet to be discussed.

Well, I do not see a need to implement and maintain two ways for doing essentially the same thing. I would recommend:
- Implementing solution 1 from [this comment](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/847#issuecomment-822949552), ___plus___
- Implementing either solution 2 or 3 from [this comment](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/847#issuecomment-822949552), or "V2" API function shown above. If asked, I would vote for the latter is it seems the simplest one.

**Comment by carlushuang** [comment]
Hi @atamazov I uploaded a draft PR implementing the "V2" API function in https://github.com/ROCmSoftwarePlatform/MIOpen/pull/883. Please take a look. That PR need a through review and decision if is missing something or if is not needed at all.
Beside, I keep the `dimsA` and `stridesA` to `size_t`, for I think both matches our internal `miopen::TensorDescriptor` `lens`, `strides` member type.

**Comment by atamazov** [comment]
[Side note] I have a feeling that we are spending too much energy on this problem, which does not seem to be a priority right now.

**Comment by carlushuang** [comment]
@atamazov OK. I just want to make some progress on this topic. But it is OK that first align the priority/decision and then move forward.

**Comment by jerryyin** [comment]
For 2, since we are about to invent a new API, why not promote it all the way up to `uint64_t`? As far as I understand, `size_t` could still be 32 bit integer that isn't much better than int.

For 3, will add a tensor format information hinder our ability to do some flexible layouts? The implication is that if we don't define a format, client will not be able to use it. I know that `TensorFlow` side has a longer term optimization plan about using MIOpen do the `YXCK` or `HWIO` format. Those are viable through MLIR binary backend, but not the typicall `NCHW` or `NHWC` layout.

**Comment by atamazov** [comment]
`size_t` is big enough to contain the size of the biggest object the system can handle.

**Comment by jerryyin** [comment]
> size of the biggest object the system can handle.

That still make it machine dependent (size_t on 32bit OS is 32 bit). What if a client has a 32 bit OS and at the same time a lot of gpu resources? He will can still overflow strides. `unit64_t` make it more deterministic to be 64 bit.

**Comment by jerryyin** [comment]
Actually from MLIR point of view, all of strides, dimensions are defined in `int64_t`, so it is the same with tensorflow. I'd prefer to be `int64_t` just so that it is consistent with both.

**Comment by atamazov** [comment]
> That still make it machine dependent (size_t on 32bit OS is 32 bit). What if a client has a 32 bit OS and at the same time a lot of gpu resources? He will can still overflow strides. `unit64_t` make it more deterministic to be 64 bit.

No, because in this case it is impossible to create a tensor > 4G on the host, so the stride can't exceed it.

**Comment by ppanchad-amd** [comment]
@atamazov Is this ticket still relevant? Thanks!

**Comment by atamazov** [comment]
@ppanchad-amd Yes. To be resolved by #2838
